Features/prune full2

liuhaox1 opened the pull request.

darren0718 committed:
1. implements checkpoint calculation by both accountDB and chainDB
2. get all groups by latest accountDB

darren0718 committed:
fork from chain-validation

darren0718 committed:
modify get history data

liuhaox1 committed:
crop develop

liuhaox1 committed:
memmax clear

liuhaox1 committed:
Merge branch 'features/get-history-data' into features/crop_trie

darren0718 committed:
implements prune command

darren0718 committed:
Merge branch 'features/crop_trie' into features/prune-full

darren0718 committed:
add fastcache

darren0718 committed:
finish full prune

darren0718 committed:
fix node

liuhaox1 committed:
crop2 dev

liuhaox1 committed:
bug fixed

liuhaox1 committed:
add shut down hook

liuhaox1 committed:
modify node cache

liuhaox1 committed:
modify queue

liuhaox1 committed:
remove no used import

liuhaox1 committed:
add trie buffer

liuhaox1 committed:
add cost log

liuhaox1 committed:
add cost log

liuhaox1 committed:
modify log show

liuhaox1 committed:
modify test

darren0718 committed:
implements fast replayer

liuhaox1 committed:
modify clear lock

darren0718 committed:
implements state verify

darren0718 committed:
Merge branch 'features/prune-full' of github.com:zvchain/zvchain into features/prune-full

liuhaox1 committed:
add log

liuhaox1 committed:
modify log

darren0718 committed:
add trie cache and meter

darren0718 committed:
Merge remote-tracking branch 'origin/features/prune-full' into features/prune-full

darren0718 committed:
Merge branch 'features/prune-full' into features/trie-verify-integrity

# Conflicts:
#	log/init.go

liuhaox1 committed:
add missing log

darren0718 committed:
verify integrity

darren0718 committed:
Merge branch 'develop' of github.com:zvchain/zvchain into features/trie-verify-integrity

liuhaox1 committed:
Merge branch 'develop' into features/trie-verify-integrity

liuhaox1 committed:
modify missing error

liuhaox1 committed:
Merge branch 'features/prune-full' into features/trie-verify-integrity

liuhaox1 committed:
Merge branch 'features/crop_trie2' into features/prune-full2

# Conflicts:
#	core/blockchain_store.go
#	storage/trie/database.go

liuhaox1 committed:
Merge branch 'features/trie-verify-integrity' into features/prune-full2

# Conflicts:
#	storage/trie/trie_test.go

liuhaox1 committed:
change config name

liuhaox1 committed:
remove no used

liuhaox1 committed:
modify

liuhaox1 committed:
bug fixed

liuhaox1 committed:
bug fixed

darren0718 committed:
prune full

liuhaox1 committed:
modify test case

liuhaox1 committed:
add test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
bug fixed

liuhaox1 committed:
add log

liuhaox1 committed:
disable rpc invoke prune mode

darren0718 committed:
prune full and verify

darren0718 committed:
Merge remote-tracking branch 'origin/features/trie-verify-integrity' into features/trie-verify-integrity

# Conflicts:
#	storage/trie/trie.go
#	storage/trie/trie_ext_test.go
#	storage/trie/trie_test.go

liuhaox1 committed:
add disable rpc

liuhaox1 committed:
Merge branch 'features/trie-verify-integrity' into features/prune-full2

# Conflicts:
#	storage/trie/trie_ext_test.go
#	storage/trie/trie_test.go

darren0718 committed:
prune full and verify

liuhaox1 committed:
modify test_case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify output format

liuhaox1 committed:
modify test case

bird512 committed:
add checkHash

bird512 committed:
Merge branch 'features/prune-full2' of github.com:zvchain/zvchain into features/prune-full2

liuhaox1 committed:
modify test case

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

darren0718 committed:
database meter out with time

darren0718 committed:
log error when reading from disk

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

darren0718 committed:
fix some problem

darren0718 committed:
persist state cache on Close

darren0718 committed:
persist state cache on Close

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify name

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

darren0718 committed:
Merge branch 'features/trie-verify-integrity' into features/prune-full2

# Conflicts:
#	cmd/gzv/cli/gzv.go
#	core/blockchain.go
#	core/state_prune.go
#	storage/account/accountdb.go

darren0718 committed:
Merge remote-tracking branch 'origin/features/prune-full2' into features/prune-full2

# Conflicts:
#	core/blockchain.go

liuhaox1 committed:
modify shutdown hook

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

darren0718 committed:
update test

liuhaox1 committed:
modify field name

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify test case

liuhaox1 committed:
add nil check

liuhaox1 committed:
modify method name

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify field name

liuhaox1 committed:
add cost log

liuhaox1 committed:
modify test case

liuhaox1 committed:
modify pop to peek

darren0718 committed:
update test

darren0718 committed:
Merge branch 'features/prune-full2' of github.com:zvchain/zvchain into features/prune-full2

liuhaox1 committed:
reset nil develop

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

liuhaox1 committed:
modify method name

ysh0566 committed:
ctvm version fix

darren0718 committed:
fix checkpoint calclation when first scan meets missing node

darren0718 committed:
Merge remote-tracking branch 'origin/features/prune-full2' into features/prune-full2

liuhaox1 committed:
bug fixed

liuhaox1 committed:
modify log

liuhaox1 committed:
modify log

liuhaox1 committed:
check add

liuhaox1 committed:
modify test  case

liuhaox1 committed:
modify test case

liuhaox1 committed:
check small db is nil

liuhaox1 committed:
triegc not empty check

darren0718 committed:
finish traverse

liuhaox1 committed:
optimization

bird512 commented on cmd/gzv/cli/miner_rpc.go:
```diff
@@ -72,13 +72,14 @@ func startHTTP(endpoint string, apis []rpc.API, modules []string, cors []string,
 	if endpoint == "" {
 		return nil
 	}
+	isPruneMode := core.BlockChainImpl.IsPruneMode()
```
we can get the prune mode config in current package directly. do not need to add this dependence for core package

sesamegu committed:
refactor

liuhaox1 committed:
add notes

darren0718 committed:
1. add pprof
2. traverse only once for same root
3. only traverse the groupkey for group account, others are pruned

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

lavrock committed:
revert subproject version

liuhaox1 committed:
rpc bug fix

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

darren0718 committed:
use ticker instead of time.Since which is cpu intense

liuhaox1 committed:
param check

bird512 commented on core/blockchain.go:
```diff
@@ -68,19 +70,33 @@ type BlockChainConfig struct {
+	batch           tasdb.Batch
+	triegc          *prque.Prque // Priority queue mapping block numbers to tries to gc
+	stateCache      account.AccountDatabase
+	running         int32 // running must be called atomically
```
better rename to shutdowning

liuhaox1 committed:
modify config

bird512 commented on core/trie_small_db_store.go:
```diff
@@ -0,0 +1,114 @@
+func (store *smallStateStore) HasStateData() bool {
+	iter := store.db.NewIterator()
+	defer iter.Release()
+	return iter.Seek([]byte(smallDbRootDatas))
```
should check the result's prefix after Seek. Seek may return data  of key greater than or equal to the given key.

liuhaox1 committed:
check iter seek

liuhaox1 committed:
merge modify

bird512 commented on core/blockchain.go:
```diff
@@ -250,6 +308,7 @@ func initBlockChain(helper types.ConsensusHelper, minerAccount types.Account) er
 	BlockChainImpl = chain
 
 	// db cache enabled
+	iteratorNodeCacheSize := 30000
```
below if is always true

bird512 commented on core/blockchain_action.go:
```diff
@@ -262,14 +263,176 @@ func (chain *FullBlockChain) validateBlock(source string, b *types.Block) (bool,
 
+// DeleteSmallDbDataByRoots will delete root list if we reset tops
+func (chain *FullBlockChain) DeleteSmallDbDataByRoots(roots []common.Hash) {
+	if len(roots) == 0 {
```
this if check can be removed

sesamegu commented on core/blockchain.go:
```diff
@@ -418,10 +471,63 @@ func (chain *FullBlockChain) GetConsensusHelper() types.ConsensusHelper {
+		if bh.Height == 0 {
+			return bh, nil
+		}
+		_, e = chain.accountDBAt(bh.Height)
```
_, e = chain.accountDBAt(bh.Height)
this is not need

liuhaox1 committed:
remove no use code

bird512 commented on core/trie_small_db_store.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+var (
+	persistentHeight = "ph"
```
Please add some comments for these three prefixes

bird512 commented on storage/trie/database.go:
```diff
@@ -316,38 +408,99 @@ func (db *NodeDatabase) insertPreimage(hash common.Hash, preimage []byte) {
 }
 
+func (db *NodeDatabase) tryGetFromCache(hash common.Hash) []byte {
+	// reset the meter if read count > 2000000, for more accuracy statistic about recent db read
```
This comment is not match to below code

darren0718 committed:
fix checkpoint Logger nil

darren0718 committed:
Merge branch 'features/prune-full2' into features/prune-full-opti

# Conflicts:
#	storage/trie/database.go

darren0718 committed:
CPU cores and max open files can be set

liuhaox1 committed:
modify small db params

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

bird512 commented on storage/trie/database.go:
```diff
@@ -701,12 +932,13 @@ func (db *NodeDatabase) Size() (common.StorageSize, common.StorageSize) {
 }
 
-// verifyIntegrity is a debug method to iterate over the entire trie stored in
+// traverse is a debug method to iterate over the entire trie stored in
```
comments  traverse not match the method name verifyIntegrity

bird512 commented on storage/trie/database.go:
```diff
@@ -747,3 +979,34 @@ func (db *NodeDatabase) accumulate(hash common.Hash, reachable map[common.Hash]s
 	}
 }
+
+func (db *NodeDatabase) printPath(root common.Hash, height uint64, rets []common.Hash) {
```
This function is never used, should add some comments for it

sesamegu committed:
refactor

darren0718 committed:
prune test

darren0718 committed:
Merge remote-tracking branch 'origin/features/prune-full2' into features/prune-full2

darren0718 committed:
add test for group data pruned

liuhaox1 committed:
cancel db deleted  compensation scheme

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

# Conflicts:
#	core/blockchain_action.go

liuhaox1 committed:
modify test case

liuhaox1 committed:
put checkpoint method to prune mode

bird512 commented on core/blockchain.go:
```diff
@@ -418,10 +499,62 @@ func (chain *FullBlockChain) GetConsensusHelper() types.ConsensusHelper {
+			return nil, fmt.Errorf("find block hash not exists,block hash is %v", preHash)
+		}
+		if bh.Height == 0 {
+			return bh, nil
```
should check the value of cnt before return directly

darren0718 committed:
add checkpoint compare test

darren0718 committed:
add checkpoint compare test

bird512 commented on storage/trie/database.go:
```diff
@@ -294,7 +392,9 @@ func (db *NodeDatabase) insert(hash common.Hash, blob []byte, node node) {
 	db.nodes[hash] = entry
-
+	if db.enablePrune {
+		db.commitFullNodes = append(db.commitFullNodes, &storeBlob{Key: hash, Raw: entry.rlp()})
```
do not need to rlp it again. can use copy()

liuhaox1 committed:
use bytes copy instead rlp

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2

liuhaox1 committed:
reset near modify

bird512 commented on core/blockchain_action.go:
```diff
@@ -262,14 +263,169 @@ func (chain *FullBlockChain) validateBlock(source string, b *types.Block) (bool,
+		return nil
+	// if big db is deleted,but small db not be deleted,we not support this stage!
+	}else if top == nil && lastStateHeight > 0{
+		return  fmt.Errorf("db is damaged,suggest delete d_mall and try again")
```
Should print this error message to console

liuhaox1 committed:
damaged print console

liuhaox1 committed:
modify panic to error

liuhaox1 committed:
modify default to 512

liuhaox1 committed:
set default param

darren0718 committed:
fix unit test

darren0718 committed:
Merge remote-tracking branch 'origin/features/prune-full2' into features/prune-full2

liuhaox1 committed:
modify test case

liuhaox1 committed:
Merge branch 'features/prune-full2' of https://github.com/zvchain/zvchain into features/prune-full2
